tree/ntuple/v7/src/RFieldMeta.cxx

```diff
@@ -198,27 +229,49 @@ void ROOT::Experimental::RClassField::ReadInClusterImpl(RNTupleLocalIndex localI
    }
 }
 
+void ROOT::Experimental::RClassField::AddReadCallbacksFromIORule(const TSchemaRule *rule)
+{
+   auto func = rule->GetReadFunctionPointer();
+   R__ASSERT(func != nullptr);
+   fReadCallbacks.emplace_back([func, classp = fClass](void *target) {
+      TVirtualObject oldObj{nullptr};
+      oldObj.fClass = classp;
+      oldObj.fObject = target;
+      func(static_cast<char *>(target), &oldObj);
+      oldObj.fClass = nullptr; // TVirtualObject does not own the value
+   });
+}
+
 void ROOT::Experimental::RClassField::BeforeConnectPageSource(Internal::RPageSource &pageSource)
 {
-   // This can happen for added base classes or non-simple members.
-   if (GetOnDiskId() == kInvalidDescriptorId) {
-      return;
-   }
-   // Gather all known sub fields in the descriptor.
+   std::vector<const TSchemaRule *> rules;
    std::unordered_set<std::string> knownSubFields;
-   {
+
+   if (GetOnDiskId() == kInvalidDescriptorId) {
+      // This can happen for added base classes or added members of class type
+      rules = FindRules(nullptr);
+   } else {
       const auto descriptorGuard = pageSource.GetSharedDescriptorGuard();
       const RNTupleDescriptor &desc = descriptorGuard.GetRef();
       const auto &fieldDesc = desc.GetFieldDescriptor(GetOnDiskId());
+
       // Check that we have the same type.
-      if (GetTypeName() != fieldDesc.GetTypeName())
+      // TODO(jblomer): relax for class rename rule
+      if (GetTypeName() != fieldDesc.GetTypeName()) {
          throw RException(R__FAIL("incompatible type name for field " + GetFieldName() + ": " + GetTypeName() +
                                   " vs. " + fieldDesc.GetTypeName()));
+      }
 
       for (auto linkId : fieldDesc.GetLinkIds()) {
          const auto &subFieldDesc = desc.GetFieldDescriptor(linkId);
          knownSubFields.insert(subFieldDesc.GetFieldName());
       }
+
+      rules = FindRules(&fieldDesc);
+   }
+
+   for (const auto rule : rules) {
+      AddReadCallbacksFromIORule(rule);
    }
 
    // Iterate over all sub fields in memory and mark those as missing that are not in the descriptor.
```

Comment:
Not sure if this is important, but before the refactoring `GetOnDiskId() == kInvalidDescriptorId` would early-exit while now we land here. I think there's no observable difference because `SetArtificial()` applies recursively to all sub-fields, so they should have already been marked by the parent...

Reply:
Isn't the point that if `GetOnDiskId() == kInvalidDescriptorId` then `knownSubFields` is empty, so the loop does nothing?

Reply:
Yes, `knownSubFields` is empty so `knownSubFields.count(field->GetFieldName())` will always be 0, marking all sub-fields as artificial. As I said, this may be fine...